src/awesome/button.rs

```diff
@@ -0,0 +1,343 @@
+    button: xcb_button_t,
+    modifiers: KeyMod
+}
+
+#[derive(Clone, Debug)]
+pub struct Button<'lua>(Table<'lua>);
+
+impl <'lua> Button<'lua> {
+    pub fn button(&self) -> rlua::Result<Value<'lua>> {
+        let button = self.0.get::<_, ButtonState>("data")?;
+        Ok(Value::Integer(button.button as _))
+    }
+
+    pub fn set_button(&self, new_val: xcb_button_t) -> rlua::Result<()> {
+        let mut button = self.0.get::<_, ButtonState>("data")?;
+        button.button = new_val;
+        self.0.set("data", button)?;
+        Ok(())
+    }
+
+    pub fn modifiers(&self) -> rlua::Result<KeyMod> {
+        let button = self.0.get::<_, ButtonState>("data")?;
+        Ok(button.modifiers)
+    }
+
+    pub fn set_modifiers(&self, mods: Table<'lua>) -> rlua::Result<()> {
+        use ::lua::mods_to_rust;
+        let mut button = self.0.get::<_, ButtonState>("data")?;
+        button.modifiers = mods_to_rust(mods)?;
+        self.0.set("data", button)?;
+        Ok(())
+    }
+}
+
+impl <'lua> ToLua<'lua> for Button<'lua> {
+    fn to_lua(self, lua: &'lua Lua) -> rlua::Result<Value<'lua>> {
+        self.0.to_lua(lua)
+    }
+}
+
+impl <'lua> Objectable<'lua, Button<'lua>, ButtonState> for Button<'lua> {
+    fn _wrap(table: Table<'lua>) -> Button {
+        Button(table)
+    }
+
+    fn get_table(self) -> Table<'lua> {
+        self.0
+    }
+}
+
+impl Display for ButtonState {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "Button: {:p}", self)
+    }
+}
+
+impl Default for ButtonState {
+    fn default() -> Self {
+        ButtonState {
+            button: xcb_button_t::default(),
+            modifiers: KeyMod::empty()
+        }
+    }
+}
+
+impl UserData for ButtonState {}
+
+/// Makes a new button stored in a table beside its signals
+pub fn allocator(lua: &Lua) -> rlua::Result<Object> {
+    let meta = lua.create_table();
+    let class = class::button_class(lua)?;
+    Ok(Button::new(lua, class)?
+       .add_to_meta(meta)?
+       .build())
+}
+
+pub fn new<'lua>(lua: &'lua Lua, _table: Table<'lua>)
+                 -> rlua::Result<Object<'lua>> {
+    allocator(lua)
+}
+
+
+pub fn init(lua: &Lua) -> rlua::Result<Class> {
+    Class::new(lua, Some(Rc::new(allocator)), None, None)?
+        .method("__call".into(), lua.create_function(new))?
+        .property(Property::new("button".into(),
+                                Some(lua.create_function(set_button)),
+                                Some(lua.create_function(get_button)),
+                                Some(lua.create_function(set_button))))?
+        .property(Property::new("modifiers".into(),
+                                Some(lua.create_function(set_modifiers)),
+                                Some(lua.create_function(get_modifiers)),
+                                Some(lua.create_function(set_modifiers))))?
+        .save_class("button")?
+        .build()
+}
+
+fn set_button<'lua>(lua: &'lua Lua, (table, val): (Table, Value))
```

Comment:
These still seem messy, due to the explicit conversion. I can't pass in an `Into<Button>`?

Reply:
Nah it's fine, that's all Lua handling code. The raw setting of the state is behind `Button` which is the important part.